File: tools/gizmo/src/runner.ts
declare const require: any;
const child_process = require('child_process');

import { ensureManifestValid } from './manifest';
import { GIZMO_COMMAND_PLAN_FORMAT, GizmoCommandPlan, GizmoManifest } from './types';

const placeholderPattern = /\{([A-Za-z_][A-Za-z0-9_-]*)\}/g;
const safeArgPattern = /^[^\0\r\n\"&|<>^%]+$/;

function unique(values: string[]): string[] {
  return Array.from(new Set(values)).sort();
}

function placeholders(template: string): string[] {
  const found: string[] = [];
  let match: RegExpExecArray | null;
  placeholderPattern.lastIndex = 0;
  while ((match = placeholderPattern.exec(template)) !== null) found.push(match[1]);
  return unique(found);
}

function quoteArg(value: string): string {
  if (!safeArgPattern.test(value)) throw new Error(`unsafe command argument: ${value}`);
  return `"${value}"`;
}

function renderTemplate(template: string, args: Record<string, string>): string {
  return template.replace(placeholderPattern, (_full, name: string) => quoteArg(args[name]));
}

export function parseArgPairs(values: string[]): Record<string, string> {
  const out: Record<string, string> = {};
  for (const item of values) {
    const index = item.indexOf('=');
    if (index <= 0) throw new Error(`expected --arg name=value, got: ${item}`);
    const name = item.slice(0, index);
    const value = item.slice(index + 1);
    if (!/^[A-Za-z_][A-Za-z0-9_-]*$/.test(name)) throw new Error(`unsafe argument name: ${name}`);
    if (value.length === 0) throw new Error(`empty argument value for ${name}`);
    out[name] = value;
  }
  return out;
}

export function buildGadgetCommandPlan(manifestInput: unknown, gadgetName: string, commandName: string, args: Record<string, string>, execute: boolean): GizmoCommandPlan {
  const manifest = ensureManifestValid(manifestInput) as GizmoManifest;
  const gadget = manifest.gadgets[gadgetName];
  if (!gadget) throw new Error(`unknown gadget: ${gadgetName}`);
  const template = gadget.commands[commandName];
  if (!template) throw new Error(`unknown command for gadget ${gadgetName}: ${commandName}`);

  const required = placeholders(template);
  const supplied = Object.keys(args).sort();
  const missing = required.filter(name => args[name] === undefined);
  const unused = supplied.filter(name => !required.includes(name));
  if (missing.length > 0) throw new Error(`missing command args: ${missing.join(', ')}`);
  if (unused.length > 0) throw new Error(`unused command args: ${unused.join(', ')}`);

  return {
    format: GIZMO_COMMAND_PLAN_FORMAT,
    gizmo: manifest.name,
    scope: 'gadget',
    name: gadgetName,
    command: commandName,
    template,
    rendered: renderTemplate(template, args),
    args,
    cwd: '.',
    execute,
    missing_args: missing,
    unused_args: unused,
  };
}

export function executeCommandPlan(plan: GizmoCommandPlan): number {
  if (!plan.execute) return 0;
  const proc = child_process.spawnSync(plan.rendered, {
    cwd: plan.cwd,
    shell: true,
    stdio: 'inherit',
  });
  if (typeof proc.status === 'number') return proc.status;
  return proc.error ? 1 : 0;
}

File: projects/lambda-script/src/lambda-script/lib/LambdaScript/Gizmo/Runner.hs
module LambdaScript.Gizmo.Runner where

import qualified Data.Aeson as Aeson
import qualified Data.Map.Strict as Map
import qualified Data.Set as Set
import Data.Char
import Data.List
import System.Exit
import System.Process

import LambdaScript.Gizmo.Manifest
import qualified LambdaScript.Gizmo.Types as GizmoTypes

parseArgPairs :: [String] -> Map.Map String String
parseArgPairs values =
  Map.fromList (map parseArgPair values)
  where
    parseArgPair item =
      case break (== '=') item of
        (_, []) -> error ("expected --arg name=value, got: " ++ item)
        (name, '=' : value)
          | null name -> error ("expected --arg name=value, got: " ++ item)
          | not (isSafeArgName name) -> error ("unsafe argument name: " ++ name)
          | null value -> error ("empty argument value for " ++ name)
          | otherwise -> (name, value)
        _ -> error ("expected --arg name=value, got: " ++ item)

    isSafeArgName :: String -> Bool
    isSafeArgName name =
      case name of
        c : rest -> (isAlpha c || c == '_') && all isArgNameChar rest
        _ -> False

    isArgNameChar c = isAlphaNum c || c == '_' || c == '-'

buildGadgetCommandPlan ::
  Aeson.Value ->
  String ->
  String ->
  Map.Map String String ->
  Bool ->
  GizmoTypes.GizmoCommandPlan
buildGadgetCommandPlan manifestInput gadgetName commandName args execute =
  let manifest = ensureManifestValid manifestInput
      gadget =
        case Map.lookup gadgetName (GizmoTypes.gizmoManifestGadgets manifest) of
          Just found -> found
          Nothing -> error ("unknown gadget: " ++ gadgetName)
      template =
        case Map.lookup commandName (GizmoTypes.gadgetManifestCommands gadget) of
          Just found -> found
          Nothing -> error ("unknown command for gadget " ++ gadgetName ++ ": " ++ commandName)
      required = placeholders template
      supplied = Set.toAscList (Map.keysSet args)
      missing = filter (\name -> not (Map.member name args)) required
      unused = filter (`notElem` required) supplied
   in if not (null missing)
        then error ("missing command args: " ++ intercalate ", " missing)
        else if not (null unused)
          then error ("unused command args: " ++ intercalate ", " unused)
          else
            GizmoTypes.GizmoCommandPlan
              { GizmoTypes.gizmoCommandPlanFormat = GizmoTypes.gizmoCommandPlanFormatTag
              , GizmoTypes.gizmoCommandPlanGizmo = GizmoTypes.gizmoManifestName manifest
              , GizmoTypes.gizmoCommandPlanScope = "gadget"
              , GizmoTypes.gizmoCommandPlanName = gadgetName
              , GizmoTypes.gizmoCommandPlanCommand = commandName
              , GizmoTypes.gizmoCommandPlanTemplate = template
              , GizmoTypes.gizmoCommandPlanRendered = renderTemplate template args
              , GizmoTypes.gizmoCommandPlanArgs = args
              , GizmoTypes.gizmoCommandPlanCwd = "."
              , GizmoTypes.gizmoCommandPlanExecute = execute
              , GizmoTypes.gizmoCommandPlanMissingArgs = missing
              , GizmoTypes.gizmoCommandPlanUnusedArgs = unused
              }

executeCommandPlan :: GizmoTypes.GizmoCommandPlan -> IO Int
executeCommandPlan plan
  | not (GizmoTypes.gizmoCommandPlanExecute plan) = return 0
  | otherwise = do
      (_, _, _, processHandle) <-
        createProcess
          (shell (GizmoTypes.gizmoCommandPlanRendered plan))
            { cwd = Just (GizmoTypes.gizmoCommandPlanCwd plan)
            , std_in = Inherit
            , std_out = Inherit
            , std_err = Inherit
            }
      status <- waitForProcess processHandle
      return
        ( case status of
            ExitSuccess -> 0
            ExitFailure code -> code
        )

placeholders :: String -> [String]
placeholders template = Set.toAscList (Set.fromList (collectPlaceholders template))

collectPlaceholders :: String -> [String]
collectPlaceholders [] = []
collectPlaceholders ('{' : rest) =
  case parsePlaceholderName rest of
    Just (name, '}' : after) -> name : collectPlaceholders after
    _ -> collectPlaceholders rest
collectPlaceholders (_ : rest) = collectPlaceholders rest

parsePlaceholderName :: String -> Maybe (String, String)
parsePlaceholderName string =
  case string of
    c : rest
      | isAlpha c || c == '_' ->
          let (name, remainder) = span isPlaceholderChar string
           in if null name then Nothing else Just (name, remainder)
    _ -> Nothing

isPlaceholderChar c = isAlphaNum c || c == '_' || c == '-'

renderTemplate :: String -> Map.Map String String -> String
renderTemplate template args = go template
  where
    go [] = []
    go ('{' : rest) =
      case parsePlaceholderName rest of
        Just (name, '}' : after) ->
          quoteArg (Map.findWithDefault "" name args) ++ go after
        _ -> '{' : go rest
    go (c : rest) = c : go rest

quoteArg :: String -> String
quoteArg value =
  if safeArgPattern value
    then "\"" ++ value ++ "\""
    else error ("unsafe command argument: " ++ value)

safeArgPattern :: String -> Bool
safeArgPattern = all isSafeArgChar
  where
    isSafeArgChar c =
      c /= '\0'
        && c /= '\r'
        && c /= '\n'
        && c /= '"'
        && c /= '&'
        && c /= '|'
        && c /= '<'
        && c /= '>'
        && c /= '^'
        && c /= '%'

